SO-2901: Fix module dependency duplicates

zoliMolnar opened the pull request:
This pull request enables updating existing module dependency refset members and adding only the new ones.

Commit:
[fix] Removed module dependency logic from the precommit state

Commit:
[fix] changed so the class returns a map of moduleId, refCompId, uuId

Commit:
[fix] changed logic so module dependency refset members are only 

created if they were not already members of said refset, otherwise
update them

Commit:
Added support for module concept creation.

Commit:
[fix] changed form multimap to map, and removed prints

Commit:
[fix] changed form multimap to map, added update logic so existing 

members will be updated instead of getting added with a different
effective time.

Commit:
[fix] added tests for module dependency

Commit:
[fix] added xtext dependecy to manifest

Commit:
[fix] added tests class and moved getBus to superclass as protected

Commit:
[fix] removed comments

Commit:
[fix] fixed rest url

nagyo commented on snomed/com.b2international.snowowl.snomed.datastore/src/com/b2international/snowowl/snomed/datastore/SnomedRefSetEditingContext.java:
```diff
@@ -717,13 +716,6 @@ public SnomedRefSet findRefSetByIdentifierConceptId(final String identifierConce
-		 */
-		SnomedModuleDependencyRefSetService dependencyRefSetService = new SnomedModuleDependencyRefSetService();
-		dependencyRefSetService.updateModuleDependenciesDuringPreCommit(getTransaction());
 	}
```
Please remove the empty method!

nagyo commented on snomed/com.b2international.snowowl.snomed.datastore.server/src/com/b2international/snowowl/datastore/server/snomed/version/SnomedPublishManager.java:
```diff
@@ -19,12 +19,15 @@
 import java.io.IOException;
-import java.util.Collection;
 import java.util.Date;
+import java.util.List;
```
Unused import

nagyo commented on snomed/com.b2international.snowowl.snomed.datastore.server/src/com/b2international/snowowl/datastore/server/snomed/version/SnomedPublishManager.java:
```diff
@@ -57,11 +60,16 @@
 import com.b2international.snowowl.terminologymetadata.CodeSystemVersion;
 import com.google.common.base.Function;
+import com.google.common.base.Strings;
+import com.google.common.collect.ListMultimap;
```
Unused import

nagyo commented on snomed/com.b2international.snowowl.snomed.datastore.server/src/com/b2international/snowowl/datastore/server/snomed/SnomedModuleDependencyCollectorService.java:
```diff
@@ -405,6 +412,18 @@ private LongKeyLongMap getConceptModuleMapping() {
+
+		for (SnomedModuleDependencyRefSetMember member : getMembers()) {
+			final Pair<String, String> pair = Tuples.pair(member.getModuleId(), member.getReferencedComponentId());
+			System.err.println(member.getUuid());
```
Please remove the syserr!

nagyo commented on snomed/com.b2international.snowowl.snomed.datastore.server/src/com/b2international/snowowl/datastore/server/snomed/version/SnomedPublishManager.java:
```diff
@@ -234,12 +244,44 @@ private boolean isCoreComponent(final EClass eClass) {
+	private void updateExistingMembers(Date effectiveTime, final SnomedRegularRefSet moduleDependencyRefSet) {
+		for (SnomedRefSetMember refSetmember : moduleDependencyRefSet.getMembers()) {
+			if (refSetmember instanceof SnomedModuleDependencyRefSetMember) {
+				final SnomedModuleDependencyRefSetMember dependecyMember = (SnomedModuleDependencyRefSetMember) refSetmember;
```
Typo!

nagyo commented on snomed/com.b2international.snowowl.snomed.api.rest.tests/src/com/b2international/snowowl/snomed/api/rest/moduledependency/SnomedModuleDependencyApiTest.java:
```diff
@@ -0,0 +1,117 @@
+import com.b2international.snowowl.snomed.datastore.request.SnomedRequests;
+import com.b2international.snowowl.snomed.snomedrefset.SnomedRefSetType;
+
+public class SnomedModuleDependencyApiTest extends AbstractSnomedApiTest {
```
Please rename this class to something like SnomedModuleDependencyRefsetTest. There is no such thing as module dependency API!

nagyo commented on snomed/com.b2international.snowowl.snomed.api.rest.tests/src/com/b2international/snowowl/snomed/api/rest/moduledependency/SnomedModuleDependencyApiTest.java:
```diff
@@ -0,0 +1,117 @@
+
+public class SnomedModuleDependencyApiTest extends AbstractSnomedApiTest {
+
+	private static final String MODULE_DEPENDECY_REFSET_ID = "900000000000534007";
```
This constant is already available at com.b2international.snowowl.snomed.SnomedConstants.Concepts.REFSET_MODULE_DEPENDENCY_TYPE

nagyo replied:
And typo!

nagyo commented on snomed/com.b2international.snowowl.snomed.api.rest.tests/src/com/b2international/snowowl/snomed/api/rest/moduledependency/SnomedModuleDependencyApiTest.java:
```diff
@@ -0,0 +1,117 @@
+
+	private static final String MODULE_DEPENDECY_REFSET_ID = "900000000000534007";
+	private static final String NORWEGIAN_MODULE_CONCEPT_ID = "51000202101";
+	private static final String ROOT_CONCEPT = "138875005";
```
com.b2international.snowowl.snomed.SnomedConstants.Concepts.ROOT_CONCEPT

nagyo commented on snomed/com.b2international.snowowl.snomed.api.rest.tests/src/com/b2international/snowowl/snomed/api/rest/moduledependency/SnomedModuleDependencyApiTest.java:
```diff
@@ -0,0 +1,117 @@
+package com.b2international.snowowl.snomed.api.rest.moduledependency;
```
A better package could be under com.b2international.snowowl.snomed.api.rest.components

nagyo commented on snomed/com.b2international.snowowl.snomed.api.rest.tests/src/com/b2international/snowowl/snomed/api/rest/moduledependency/SnomedModuleDependencyApiTest.java:
```diff
@@ -0,0 +1,117 @@
+	private static final String ROOT_CONCEPT = "138875005";
+
+	@Test
+	public void testForModuleDependencyMembers() {
```
Please change the method name to something more meaningful. I don't really have a clue about what's the goal of the test just by reading it's name.

nagyo commented on snomed/com.b2international.snowowl.snomed.api.rest.tests/src/com/b2international/snowowl/snomed/api/rest/moduledependency/SnomedModuleDependencyApiTest.java:
```diff
@@ -0,0 +1,117 @@
+		final String shortName = "SNOMEDCT-MODULEDEPENDENCY";
+		createCodeSystem(branchPath, shortName).statusCode(201);
+		
+		// get number of module_dependecy_refset_member
```
Typo

nagyo commented on snomed/com.b2international.snowowl.snomed.api.rest.tests/src/com/b2international/snowowl/snomed/api/rest/moduledependency/SnomedModuleDependencyApiTest.java:
```diff
@@ -0,0 +1,117 @@
+		
+		// get number of module_dependecy_refset_member
+
+		int sizeBeforeConceptCreation = SnomedRequests.prepareSearchRefSet()
```
SnomedRequests.prepareSearchMember() would be a better choice here! (and then you could use filterByRefset(String id) )

nagyo commented on snomed/com.b2international.snowowl.snomed.api.rest.tests/src/com/b2international/snowowl/snomed/api/rest/moduledependency/SnomedModuleDependencyApiTest.java:
```diff
@@ -0,0 +1,117 @@
+		createVersion(shortName, versionId, effectiveDate).statusCode(201);
+		getVersion(shortName, versionId).statusCode(200);
+
+		// get number of module_dependecy_refset_member after versioning
```
typo

nagyo commented on snomed/com.b2international.snowowl.snomed.api.rest.tests/src/com/b2international/snowowl/snomed/api/rest/moduledependency/SnomedModuleDependencyApiTest.java:
```diff
@@ -0,0 +1,117 @@
+		getVersion(shortName, versionId).statusCode(200);
+
+		// get number of module_dependecy_refset_member after versioning
+		List<SnomedReferenceSetMember> items = SnomedRequests.prepareSearchRefSet()
```
Use prepareSearchMember() here as well!

nagyo commented on snomed/com.b2international.snowowl.snomed.api.rest.tests/src/com/b2international/snowowl/snomed/api/rest/moduledependency/SnomedModuleDependencyApiTest.java:
```diff
@@ -0,0 +1,117 @@
+				.getSync()
+				.getItems().get(0).getMembers().getItems();
+		
+		items.forEach(i -> System.out.println(i.getModuleId()));
```
Please remove the sysout

nagyo commented on snomed/com.b2international.snowowl.snomed.api.rest.tests/src/com/b2international/snowowl/snomed/api/rest/moduledependency/SnomedModuleDependencyApiTest.java:
```diff
@@ -0,0 +1,117 @@
+		items.forEach(i -> System.out.println(i.getModuleId()));
+		int sizeAfterVersioningAndConceptCreation = items.size();
+		
+		assertNotEquals("Csocsóa$ztalt megborító megamix", sizeAfterVersioningAndConceptCreation, sizeAfterConceptCreation);
```
Please change the assertion message to something meaningful.

nagyo commented on snomed/com.b2international.snowowl.snomed.api.rest.tests/src/com/b2international/snowowl/snomed/api/rest/moduledependency/SnomedModuleDependencyApiTest.java:
```diff
@@ -0,0 +1,117 @@
+				.build();
+		createComponent(branchPath, SnomedComponentType.CONCEPT, norwegianModuleRequestBody).statusCode(201);
+		
+		Map<?,?> inferredRelationshipRequestBody = SnomedRestFixtures.createRelationshipRequestBody(NORWEGIAN_MODULE_CONCEPT_ID, Concepts.IS_A, Concepts.MODULE_ROOT, NORWEGIAN_MODULE_CONCEPT_ID, CharacteristicType.INFERRED_RELATIONSHIP, 0)
```
Please try to keep the maximum line length of 120 characters

nagyo replied:
Since it's SNOMED CT, it would be better to create a **stated** IS-A relationship instead of an inferred one or create a stated and an inferred if possible.

nagyo commented on snomed/com.b2international.snowowl.snomed.api.rest.tests/src/com/b2international/snowowl/snomed/api/rest/moduledependency/SnomedModuleDependencyApiTest.java:
```diff
@@ -0,0 +1,117 @@
+		
+		createComponent(branchPath, SnomedComponentType.RELATIONSHIP, inferredRelationshipRequestBody).statusCode(201);
+		
+		Map<?, ?> norwegianConceptRequestBody = createConceptRequestBody(ROOT_CONCEPT, NORWEGIAN_MODULE_CONCEPT_ID, SnomedApiTestConstants.UK_PREFERRED_MAP)
```
I don't really see the reason behind creating this extra concept? You have already created a concept which is in the norwegian module (the module concept itself).

nagyo commented on snomed/com.b2international.snowowl.snomed.api.rest.tests/src/com/b2international/snowowl/snomed/api/rest/moduledependency/SnomedModuleDependencyApiTest.java:
```diff
@@ -0,0 +1,117 @@
+				.getSync();
+
+		concepts.forEach(c -> assertEquals("Effective time must still be null", null, c.getEffectiveTime()));
+		int sizeAfterConceptCreation = SnomedRequests.prepareSearchRefSet()
```
Use prepareSearchMember() here as well

nagyo commented on snomed/com.b2international.snowowl.snomed.api.rest.tests/src/com/b2international/snowowl/snomed/api/rest/moduledependency/SnomedModuleDependencyApiTest.java:
```diff
@@ -0,0 +1,117 @@
+				.getSync()
+				.getItems().get(0).getMembers().getItems().size();
+
+		assertEquals("After creating module concept the moduleDependencyRefset shouldn't change", sizeBeforeConceptCreation, sizeAfterConceptCreation);
```
The message is a bit misleading. We have just created **two** concepts not just the module!

nagyo commented on snomed/com.b2international.snowowl.snomed.api.rest.tests/src/com/b2international/snowowl/snomed/api/rest/moduledependency/SnomedModuleDependencyApiTest.java:
```diff
@@ -0,0 +1,117 @@
+		items.forEach(i -> System.out.println(i.getModuleId()));
+		int sizeAfterVersioningAndConceptCreation = items.size();
+		
+		assertNotEquals("Csocsóa$ztalt megborító megamix", sizeAfterVersioningAndConceptCreation, sizeAfterConceptCreation);
```
In general it would be nice to track not just the amount of refset members but the specific module -> module pairs somehow. So we can be sure that no unwanted member was created. This approach should be followed during the whole test.

nagyo replied:
Or track the number of members where the module is INT

nagyo reviewed (changes requested):
Test passed both on the vanilla Snow Owl 6.x branch and on IHTSDO's develop branch **without** your changes in the module collector service. Please rewrite it to reproduce the original issue and then we can check the additional logic applied to those services.

General advice for writing commit messages:
- if you use the square bracket prefix try to add a meaningful word in between which can describe the context of the commit for someone who will inspect your changes in the future. In this case the word "fix" is too general, it could be something like [snomed] or [moduledependency] or [snomed.mdr] etc..
- if you are changing just a class in a commit and you want to reference it in the commit message then use the class's name instead of writing "the class". This will also help someone in the future to process your commit's goals faster! 
- instead of: "changed so the class returns a map of moduleId, refCompId, uuId" you could use: "Return a map of moduleId, refCompId and UUID instead of a member collection". The class does not return anything, the method does!

Commit:
[moduledependency] Rename test class to SnomedModuleDependencyRefsetTest

And placed into the components to the snomed.api.rest.components package

Commit:
[moduledependency] Removed unused imports, and sysout

Commit:
[moduledependency] Removed unused imports, and modified logic, so it's

not changing the effective times of existing module dependency members

Commit:
[moduledependency] Removed empty method.

Commit:
[moduledependency] removed sysout, changed prepareSearchRefSet to

prepareSearchMember, modified the test logic so it's taking effective
times into account instead of checking the size of the module dependency
refset.

Commit:
[moduledependency] Removed sysout.

nagyo commented on snomed/com.b2international.snowowl.snomed.api.rest.tests/src/com/b2international/snowowl/snomed/api/rest/components/SnomedModuleDependencyRefsetTest.java:
```diff
@@ -0,0 +1,115 @@
+public class SnomedModuleDependencyRefsetTest extends AbstractSnomedApiTest {
+
+	private static final String NORWEGIAN_MODULE_CONCEPT_ID = "51000202101";
+	
```
Please remove all extra empty lines

nagyo commented on snomed/com.b2international.snowowl.snomed.api.rest.tests/src/com/b2international/snowowl/snomed/api/rest/components/SnomedModuleDependencyRefsetTest.java:
```diff
@@ -0,0 +1,115 @@
+
+		SnomedReferenceSetMembers moduleDependencyMembers1 = SnomedRequests.prepareSearchMember()
+				.all()
+				.filterByRefSet(Concepts.REFSET_MODULE_DEPENDENCY_TYPE)
```
Filter for active members (just to be sure)

nagyo replied:
.filterByModules(INT_MODULE_IDS)
.filterByReferencedComponent(INT_MODULE_IDS)

To make sure we collect only INT module dependency members

nagyo commented on snomed/com.b2international.snowowl.snomed.api.rest.tests/src/com/b2international/snowowl/snomed/api/rest/components/SnomedModuleDependencyRefsetTest.java:
```diff
@@ -0,0 +1,115 @@
+
+		Map<Pair<String, String>, Date> moduleToReferencedComponentAndEffectiveDateMap = Maps.newHashMap();
+
+		SnomedReferenceSetMembers moduleDependencyMembers1 = SnomedRequests.prepareSearchMember()
```
Please rename this to something meaningful, like intModuleDependencyMembers

nagyo commented on snomed/com.b2international.snowowl.snomed.api.rest.tests/src/com/b2international/snowowl/snomed/api/rest/components/SnomedModuleDependencyRefsetTest.java:
```diff
@@ -0,0 +1,115 @@
+				.execute(getBus())
+				.getSync();
+
+		concepts.forEach(c -> assertEquals("Effective time must still be null", null, c.getEffectiveTime()));
```
If you check the effective time of newly created components then it would be nice to check that after versioning they have the new effective time set!

nagyo commented on snomed/com.b2international.snowowl.snomed.api.rest.tests/src/com/b2international/snowowl/snomed/api/rest/components/SnomedModuleDependencyRefsetTest.java:
```diff
@@ -0,0 +1,115 @@
+		moduleDependencyMembersAfterVersioning.getItems().forEach(member-> {
+			final Pair<String, String> pair = Tuples.pair(member.getModuleId(), member.getReferencedComponent().getId());
+			final Date originalMemberEffectiveTime = moduleToReferencedComponentAndEffectiveDateMap.get(pair);
+			final Date versionDate = EffectiveTimes.parse(effectiveDate, DateFormats.SHORT);
```
This could be moved outside of the for loop

nagyo commented on snomed/com.b2international.snowowl.snomed.api.rest.tests/src/com/b2international/snowowl/snomed/api/rest/components/SnomedModuleDependencyRefsetTest.java:
```diff
@@ -0,0 +1,115 @@
+			final Pair<String, String> pair = Tuples.pair(member.getModuleId(), member.getReferencedComponent().getId());
+			final Date originalMemberEffectiveTime = moduleToReferencedComponentAndEffectiveDateMap.get(pair);
+			final Date versionDate = EffectiveTimes.parse(effectiveDate, DateFormats.SHORT);
+			if(originalMemberEffectiveTime != null) {
```
moduleToReferencedComponentAndEffectiveDateMap.containsKey(pair) would be a better condition to check

nagyo replied:
This is not fixed

nagyo commented on snomed/com.b2international.snowowl.snomed.api.rest.tests/src/com/b2international/snowowl/snomed/api/rest/components/SnomedModuleDependencyRefsetTest.java:
```diff
@@ -0,0 +1,115 @@
+		getVersion(shortName, versionId).statusCode(200);
+
+		SnomedReferenceSetMembers moduleDependencyMembersAfterVersioning = SnomedRequests.prepareSearchMember()
+				.all()
```
filter for active members!

nagyo replied:
Still filter for active members

nagyo commented on snomed/com.b2international.snowowl.snomed.api.rest.tests/src/com/b2international/snowowl/snomed/api/rest/components/SnomedModuleDependencyRefsetTest.java:
```diff
@@ -0,0 +1,115 @@
+import java.util.Date;
+import java.util.Map;
+
+import org.eclipse.xtext.util.Pair;
```
I have a compile error here when I check out this branch

nagyo commented on snomed/com.b2international.snowowl.snomed.api.rest.tests/src/com/b2international/snowowl/snomed/api/rest/components/SnomedModuleDependencyRefsetTest.java:
```diff
@@ -0,0 +1,115 @@
+			final Pair<String, String> pair = Tuples.pair(member.getModuleId(), member.getReferencedComponent().getId());
+			final Date originalMemberEffectiveTime = moduleToReferencedComponentAndEffectiveDateMap.get(pair);
+			final Date versionDate = EffectiveTimes.parse(effectiveDate, DateFormats.SHORT);
+			if(originalMemberEffectiveTime != null) {
```
Missing space!

nagyo replied:
This is not fixed

nagyo commented on snomed/com.b2international.snowowl.snomed.api.rest.tests/src/com/b2international/snowowl/snomed/api/rest/components/SnomedModuleDependencyRefsetTest.java:
```diff
@@ -0,0 +1,115 @@
+			final Date versionDate = EffectiveTimes.parse(effectiveDate, DateFormats.SHORT);
+			if(originalMemberEffectiveTime != null) {
+				assertEquals("Effective dates on existing module dependency members shouldn't be updated after versioning" , originalMemberEffectiveTime,  member.getEffectiveTime());
+				moduleToReferencedComponentAndEffectiveDateMap.remove(pair);
```
Why do you remove the pairs here???

nagyo replied:
This is not fixed

nagyo commented on snomed/com.b2international.snowowl.snomed.datastore.server/src/com/b2international/snowowl/datastore/server/snomed/version/SnomedPublishManager.java:
```diff
@@ -234,12 +242,41 @@ private boolean isCoreComponent(final EClass eClass) {
+		if (!incomingModuleDependencyRefSetMembers.isEmpty()) {
+			removeDuplicateMembers(effectiveTime, moduleDependencyRefSet);
+			incomingModuleDependencyRefSetMembers.forEach((k, v) -> {
+					// Create snomedMoudleDependencyRefSetMember out of k+v
```
Typo

nagyo commented on snomed/com.b2international.snowowl.snomed.datastore.server/src/com/b2international/snowowl/datastore/server/snomed/version/SnomedPublishManager.java:
```diff
@@ -234,12 +242,41 @@ private boolean isCoreComponent(final EClass eClass) {
+					memberToAdd.setModuleId(k.getFirst());
+					memberToAdd.setReferencedComponentId(k.getSecond());
+					
+					adjustRelased(memberToAdd);
```
Please fix the typo in the method's name

nagyo commented on snomed/com.b2international.snowowl.snomed.api.rest.tests/src/com/b2international/snowowl/snomed/api/rest/components/SnomedModuleDependencyRefsetTest.java:
```diff
@@ -0,0 +1,115 @@
+		concepts.forEach(c -> assertEquals("Effective time must still be null", null, c.getEffectiveTime()));
+				
+		// version branch
+		final String effectiveDate = getNextAvailableEffectiveDateAsString(SnomedTerminologyComponentConstants.SNOMED_SHORT_NAME);
```
The method parameter must be the extension's short name (not the core SNOMED CT short name) !

nagyo commented on snomed/com.b2international.snowowl.snomed.api.rest.tests/src/com/b2international/snowowl/snomed/api/rest/components/SnomedModuleDependencyRefsetTest.java:
```diff
@@ -0,0 +1,115 @@
+				.execute(getBus())
+				.getSync();
+		
+		moduleDependencyMembersAfterVersioning.getItems().forEach(member-> {
```
getItems() is not necessary here

nagyo commented on snomed/com.b2international.snowowl.snomed.api.rest.tests/src/com/b2international/snowowl/snomed/api/rest/components/SnomedModuleDependencyRefsetTest.java:
```diff
@@ -0,0 +1,115 @@
+				.execute(getBus())
+				.getSync();
+
+		moduleDependencyMembers1.getItems().forEach(member -> {
```
getItems() is not necessary here

nagyo commented on snomed/com.b2international.snowowl.snomed.api.rest.tests/src/com/b2international/snowowl/snomed/api/rest/components/SnomedModuleDependencyRefsetTest.java:
```diff
@@ -0,0 +1,115 @@
+	
+
+	@Test
+	public void moduleDependecyDuplicateMemberTest() {
```
There is no duplicate member issue here. We are trying to avoid the update of INT members when an extension version is created.

nagyo commented on snomed/com.b2international.snowowl.snomed.api.rest.tests/src/com/b2international/snowowl/snomed/api/rest/components/SnomedModuleDependencyRefsetTest.java:
```diff
@@ -0,0 +1,115 @@
+			final Date originalMemberEffectiveTime = moduleToReferencedComponentAndEffectiveDateMap.get(pair);
+			final Date versionDate = EffectiveTimes.parse(effectiveDate, DateFormats.SHORT);
+			if(originalMemberEffectiveTime != null) {
+				assertEquals("Effective dates on existing module dependency members shouldn't be updated after versioning" , originalMemberEffectiveTime,  member.getEffectiveTime());
```
Effective time shouldn't be updated for INT members (where both the moduleId and the referenced component id is an international module). This is not equivalent with what you wrote here!

nagyo replied:
This is not fixed

nagyo committed:
[snomed.tests] Add test case for updating existing module dependency...

...members

nagyo reviewed (changes requested):
Please fix the changes requested and revisit the logic you changed in SnomedPublishManager. I'll add a test case that is responsible for checking if an existing members is correctly updated or not. This test does not pass with your current changes.

Commit:
[moduledependency] Moved and changed the logic of removeDuplicateMembers

method, changed incomingModuleDependencyRefSetMembers to
changedModuleDependencyRefSetMembers, and changed it's type from Map to
Collection.

Commit:
[moduledependency] Changed type of members.

Commit:
[moduledependency] Added xtext.util as imported package to manifest.

Commit:
Revert "[moduledependency] Added xtext.util as imported package to manifest."

This reverts commit 03440bd22ec96b3c177c5e05e25b900cdf8f4f4e.

Commit:
[moduledependency] Added xtext.util as imported package to manifest.

Commit:
[moduledependency] Changed return type of collectModuleMembers to...

	-Collection modified logic of initConfiguration so at start existing
		module members won't be always added to the config.
	 -Renamed tryCreateMember and changed its logic so it doesn't create
		new members but rather adds a pair of sourceModuleId and
		targetModuleId.
	 -Removed Unused Methods

Commit:
[moduledependency] Made sure to get only intDependencyMembers by...

filtering for int module id-s in the request.

Added effective time assertation for the norwegian module concept after
versioning.

Moved effectiveDate outside of loop.

nagyo commented on snomed/com.b2international.snowowl.snomed.api.rest.tests/src/com/b2international/snowowl/snomed/api/rest/components/SnomedModuleDependencyRefsetTest.java:
```diff
@@ -0,0 +1,185 @@
+import com.google.common.collect.ImmutableSet;
+import com.google.common.collect.Maps;
+
+@FixMethodOrder(MethodSorters.NAME_ASCENDING)
```
I don't see why is this needed?

nagyo commented on snomed/com.b2international.snowowl.snomed.api.rest.tests/src/com/b2international/snowowl/snomed/api/rest/components/SnomedModuleDependencyRefsetTest.java:
```diff
@@ -81,29 +86,33 @@ public void updateExistingModuleDependencyMembers() {
 	
 	@Test
-	public void moduleDependecyDuplicateMemberTest() {
+	public void moduleDependecyMemberEffectiveTimeUpdateTest() {
```
Typo

nagyo commented on snomed/com.b2international.snowowl.snomed.api.rest.tests/src/com/b2international/snowowl/snomed/api/rest/components/SnomedModuleDependencyRefsetTest.java:
```diff
@@ -0,0 +1,185 @@
+		// create both inferred and stated relationships
+		Map<?, ?> inferredRelationshipRequestBody = SnomedRestFixtures
+				.createRelationshipRequestBody(NORWEGIAN_MODULE_CONCEPT_ID, Concepts.IS_A, Concepts.MODULE_ROOT, NORWEGIAN_MODULE_CONCEPT_ID, CharacteristicType.INFERRED_RELATIONSHIP, 0)
+				.put("commitComment", "Created inferred is_a from the norwegian rule to SCT_MODULE_CORE").build();
```
What is a norwegian **rule**?

nagyo commented on snomed/com.b2international.snowowl.snomed.api.rest.tests/src/com/b2international/snowowl/snomed/api/rest/components/SnomedModuleDependencyRefsetTest.java:
```diff
@@ -0,0 +1,185 @@
+		
+		Map<?, ?> statedRelationshipRequestBody = SnomedRestFixtures
+				.createRelationshipRequestBody(NORWEGIAN_MODULE_CONCEPT_ID, Concepts.IS_A, Concepts.MODULE_ROOT, NORWEGIAN_MODULE_CONCEPT_ID, CharacteristicType.STATED_RELATIONSHIP, 0)
+				.put("commitComment", "Created state is_a from the norwegian rule to SCT_MODULE_CORE").build();
```
Same "rule" here

nagyo commented on snomed/com.b2international.snowowl.snomed.datastore.server/src/com/b2international/snowowl/datastore/server/snomed/SnomedModuleDependencyCollectorService.java:
```diff
@@ -401,20 +358,4 @@ private LongKeyLongMap getConceptModuleMapping() {
-		return getConfiguration().getMembers();
-	}
-	
-	private Multimap<Long, Long> getModuleMapping() {
```
If this property is not used at all then please remove it from ModuleCollectorConfiguration

nagyo commented on snomed/com.b2international.snowowl.snomed.datastore.server/src/com/b2international/snowowl/datastore/server/snomed/SnomedModuleDependencyCollectorService.java:
```diff
@@ -401,20 +358,4 @@ private LongKeyLongMap getConceptModuleMapping() {
-		return getConfiguration().getModuleMapping();
-	}
-	
-	private SnomedRefSet getModuleDependencyRefSet() {
```
If this property is not used at all then please remove it from ModuleCollectorConfiguration

nagyo commented on snomed/com.b2international.snowowl.snomed.datastore.server/src/com/b2international/snowowl/datastore/server/snomed/SnomedModuleDependencyCollectorService.java:
```diff
@@ -401,20 +358,4 @@ private LongKeyLongMap getConceptModuleMapping() {
-	}
-	
-	private SnomedReferenceSetMembers getExistingModules() {
-		return getConfiguration().getExistingModules();
```
If this property is not used at all then please remove it from ModuleCollectorConfiguration

nagyo commented on snomed/com.b2international.snowowl.snomed.datastore.server/src/com/b2international/snowowl/datastore/server/snomed/version/SnomedPublishManager.java:
```diff
@@ -75,7 +81,9 @@
 	private Set<String> componentIdsToPublish = newHashSet();
-	private Collection<SnomedModuleDependencyRefSetMember> newModuleDependencyRefSetMembers;
+	
+	// Map<Pair<moduleId, referencedComponentId>, Uuid> 
```
Why is this comment here?

nagyo commented on snomed/com.b2international.snowowl.snomed.datastore.server/src/com/b2international/snowowl/datastore/server/snomed/version/SnomedPublishManager.java:
```diff
@@ -75,7 +81,9 @@
-	private Collection<SnomedModuleDependencyRefSetMember> newModuleDependencyRefSetMembers;
+	
+	// Map<Pair<moduleId, referencedComponentId>, Uuid> 
+	private Collection<Pair<String, String>> changedModuleDependencyRefSetMembers;
```
This collection contains rather module and referenced component ids than actual refset members. Please rename it to reflect the purpose better!

nagyo commented on snomed/com.b2international.snowowl.snomed.datastore.server/src/com/b2international/snowowl/datastore/server/snomed/version/SnomedPublishManager.java:
```diff
@@ -234,15 +242,43 @@ private boolean isCoreComponent(final EClass eClass) {
-			adjustEffectiveTime(member, effectiveTime);
-			moduleDependencyRefSet.getMembers().add(member);
+	private void adjustDependencyRefSetMembers(Date effectiveTime) {
+		// Update existing, add new members to moduleDependecyRefSet
```
Typo

nagyo commented on snomed/com.b2international.snowowl.snomed.datastore.server/src/com/b2international/snowowl/datastore/server/snomed/version/SnomedPublishManager.java:
```diff
@@ -234,15 +242,43 @@ private boolean isCoreComponent(final EClass eClass) {
+			final SnomedRegularRefSet moduleDependencyRefSet = getEditingContext().lookup(REFSET_MODULE_DEPENDENCY_TYPE, SnomedRegularRefSet.class);
+			changedModuleDependencyRefSetMembers.forEach(moduleToRefComp -> {
+				boolean modified = false;
+				for (SnomedRefSetMember refSetmember : moduleDependencyRefSet.getMembers()) {
```
Almost good, but we need to prepare for certain extensions who manage their module dependency members in a different way than IHTSDO does. Let's talk about it when you are making changes the next time.

nagyo commented on snomed/com.b2international.snowowl.snomed.datastore.server/src/com/b2international/snowowl/datastore/server/snomed/version/SnomedPublishManager.java:
```diff
@@ -234,15 +242,43 @@ private boolean isCoreComponent(final EClass eClass) {
+						final Pair<String, String> pair = Tuples.pair(dependencyMember.getModuleId(), dependencyMember.getReferencedComponentId());
+						
+						if (moduleToRefComp.equals(pair)) {
+							adjustRelased(dependencyMember);
```
This was not yours but please fix the typo in the method's name!

Commit:
[moduleDependency] Changed comment, renamed field. Changed logic of...

adjustDependencyrefSetMembers so it's adjusting the latest module
dependency member's effective time.

Commit:
[moduleDependency] Fixed typo in method's name, fixed typo in comments

removed removal of members, fixed assert message.

Commit:
[moduleDependency] Removed unused methods and fields from...

ModuleCollectorConfiguration class.

nagyo committed:
Merge remote-tracking branch 'origin/6.x' into
issue/fix-module-dependency-duplicates

Conflicts:
	snomed/com.b2international.snowowl.snomed.datastore/src/com/b2international/snowowl/snomed/datastore/SnomedEditingContext.java
	snomed/com.b2international.snowowl.snomed.datastore/src/com/b2international/snowowl/snomed/datastore/SnomedRefSetEditingContext.java

nagyo committed:
SO-2901: Update test method's name to reflect the purpose better

https://snowowl.atlassian.net/browse/SO-2901
